src/domain/flow-system/services/src/flow/flow_service_impl.rs

```diff
@@ -757,22 +765,49 @@ impl FlowServiceImpl {
                             matches!(compaction_rule, CompactionRule::MetadataOnly(_));
                     };
 
-                    LogicalPlan::HardCompactionDataset(HardCompactionDataset {
+                    Ok(LogicalPlan::HardCompactionDataset(HardCompactionDataset {
                         dataset_id: flow_key.dataset_id.clone(),
                         max_slice_size,
                         max_slice_records,
                         keep_metadata_only,
-                    })
+                    }))
+                }
+                DatasetFlowType::Reset => {
+                    let new_head_hash = if let Some(config_rule) = config_snapshot
+                        && let FlowConfigurationSnapshot::Reset(reset_rule) = config_rule
+                    {
+                        Some(reset_rule.new_head_hash.clone())
+                    } else {
+                        let dataset_handle = self
+                            .dataset_repo
+                            .resolve_dataset_ref(&flow_key.dataset_id.as_local_ref())
+                            .await
+                            .int_err()?;
+                        let dataset = self
+                            .dataset_repo
+                            .get_dataset(&dataset_handle.as_local_ref())
+                            .await
+                            .int_err()?;
+                        dataset
+                            .as_metadata_chain()
+                            .try_get_ref(&BlockRef::Head)
```

Comment:
This `else` confuses me.

The condition is:
```rust
if let Some(config_rule) = config_snapshot
  && let FlowConfigurationSnapshot::Reset(reset_rule) = config_rule
```
so:
- if a configuration was provided - **when wouldn't it?**
- and configuration is for the correct type of the rule - **why don't we immediately fail the flow if it isn't?**
- then we reset to the block that is already a `head` - **why do all these extra steps to perform a no-op?**

Reply:
Yeah, it is logical mistake. It should be seed block as default

Reply:
I'd recommend to move "seed block as default" logic all the way up to API layer, as it's mostly a user experience thing that we might reconsider in future - it doesn't belong in the core.

Reply:
I agree, and this would allow to revert new dependency + `async` modifier on this method.

Reply:
Sounds great, and issue with tests stuck will not be reproducable